src/Streamly/Coreutils/Uniq.hs

```diff
@@ -0,0 +1,215 @@
+            compareFunc =
+               if ignCase
+               then compareIgnCase
+               else (==)
+            compareWith :: (String -> String -> Bool) -> String -> String -> Bool
+            compareWith eq s1 s2 = eq s1 s2
+            -- ^
+            -- Gives the number of characters to be deleted
+            -- from the beginning of the string to skip @n@
+            -- non-space fields\/strings
+            --
+            -- @since 0.1.0.0
+            indexAfterSkippingNWords :: Int -> String -> Int
+            indexAfterSkippingNWords n str = go 0 n
+               where
+                  len = length str
+                  go i 0 = i
+                  go i _ | i >= (len - 1) = len
+                  go i xx = if isChar (str !! i) && isSpace (str !! (i + 1))
+                           then go (i + 1) (xx - 1)
+                           else go (i + 1) xx
+                  isChar = not . isSpace
+         in
+            case mz of
+               Just z -> compareWith
+                         compareFunc
+                         (slice (ix1 + y) z str1)
+                         (slice (ix2 + y) z str2)
+               _  -> compareWith
+                     compareFunc
+                     (drop (ix1 + y) str1)
+                     (drop (ix2 + y) str2)
+
+
+-- |
+-- Applies the comparison function to adjacent strings in the stream
+-- and returns a @(Int, String)@ stream where the integer
+-- is the count of occurrences of the string
+--
+-- @since 0.1.0.0
+getRepetition :: (IsStream t, Monad m) => (String -> String -> Bool) -> t m String -> t m (Int, String)
+getRepetition comparator = S.groupsBy comparator
+                              (FL.mkPureId (\(i, a) s ->
+                                  if a == ""
+                                  then (i + 1, s)
+                                  else (i + 1, a)) (0, ""))
+
+
+-- |
+-- Generates a @UniqResult@ stream
+-- applying uniq on the input stream
+-- according to the the @UniqOptions@ specified
+--
+-- @since 0.1.0.0
+uniq :: (IsStream t, Monad m) => UniqOptions -> t m String -> t m UniqResult
+uniq opt strm =
+    let
+        val (_, x) = Val x
+        eq n (i, _) = i == n
+        ge n (i, _) = i > n
+        cnt (n, s) = Count n s
+        convert predicate = S.map cnt
+            $ S.filter predicate
+            $ getRepetition (compareUsingOptions opt) strm
+    in
+        if count opt == True
+        then
+            case (output opt) of
+                Unique -> convert (eq 1)
+                _ -> convert (ge 1)
+        else
+            case (output opt) of
+                Unique -> S.map val
+                     $ S.filter (eq 1)
+                     $ getRepetition (compareUsingOptions opt) strm
+                Repeated -> S.map val
+                     $ S.filter (ge 1)
+                     $ getRepetition (compareUsingOptions opt) strm
+                Duplicate -> S.concatMap (\(i, x) -> S.replicate i (Val x))
```

Comment:
You don't need to use `concatMap` here. Instead add a function that converts `t m UniqResult` to `t m String`.

Reply:
why? Aren't we leaving the final result to `t m UniqResult` ?